includes/class-newspack-popups-custom-placements.php

```diff
@@ -0,0 +1,286 @@
+			true
+		);
+
+		\wp_localize_script(
+			'newspack-popups-blocks',
+			'newspack_popups_blocks_data',
+			[
+				'custom_placements' => self::get_custom_placements(),
+			]
+		);
+
+		\wp_register_style(
+			'newspack-popups-blocks',
+			plugins_url( '../dist/blocks.css', __FILE__ ),
+			[],
+			filemtime( dirname( NEWSPACK_POPUPS_PLUGIN_FILE ) . '/dist/blocks.css' )
+		);
+		wp_style_add_data( 'newspack-popups-blocks', 'rtl', 'replace' );
+		wp_enqueue_style( 'newspack-popups-blocks' );
+	}
+
+	/**
+	 * Enqueue front-end assets.
+	 */
+	public static function manage_view_assets() {
+		// Do nothing in editor environment.
+		if ( is_admin() ) {
+			return;
+		}
+
+		$src_directory  = dirname( NEWSPACK_POPUPS_PLUGIN_FILE ) . '/src/blocks/';
+		$dist_directory = dirname( NEWSPACK_POPUPS_PLUGIN_FILE ) . '/dist/';
+		$iterator       = new \DirectoryIterator( $src_directory );
+
+		foreach ( $iterator as $block_directory ) {
+			if ( ! $block_directory->isDir() || $block_directory->isDot() ) {
+				continue;
+			}
+			$type = $block_directory->getFilename();
+
+			/* If view.php is found, include it and use for block rendering. */
+			$view_php_path = $src_directory . $type . '/view.php';
+			if ( file_exists( $view_php_path ) ) {
+				include_once $view_php_path;
+				continue;
+			}
+		}
+	}
+
+	/**
+	 * Initialise REST API endpoints.
+	 */
+	public static function rest_api_init() {
+		\register_rest_route(
+			'newspack-popups/v1',
+			'custom-placement',
+			[
+				'methods'             => \WP_REST_Server::READABLE,
+				'callback'            => [ __CLASS__, 'api_get_prompts_for_custom_placement' ],
+				'permission_callback' => '__return_true',
+			]
+		);
+	}
+
+	/**
+	 * Get prompts assigned to a given custom placement.
+	 *
+	 * @param WP_REST_Request $request Request object.
+	 * @return WP_REST_Response Prompt IDs and titles matching the request params.
+	 */
+	public static function api_get_prompts_for_custom_placement( $request ) {
+		$custom_placement_id = self::validate_custom_placement_id( $request['custom_placement'] );
+
+		if ( empty( $custom_placement_id ) ) {
+			return new \WP_Error(
+				'newspack_popups_custom_placement',
+				esc_html__( 'Invalid custom placement ID.', 'newspack-popups' ),
+				[
+					'status' => 400,
+					'level'  => 'fatal',
+				]
+			);
+		}
+
+		$prompts = self::get_prompts_for_custom_placement( [ $custom_placement_id ] );
+
+		if ( ! empty( $prompts ) ) {
+			$sorted_prompts = array_map(
+				function( $post ) {
+					$segment_ids = get_post_meta( $post->ID, 'selected_segment_id', true );
+					$segments    = [
+						[
+							'name'     => 'Everyone',
+							'priority' => PHP_INT_MAX,
+						],
+					];
+
+					if ( $segment_ids ) {
+						$segment_ids = explode( ',', $segment_ids );
+						$segments    = array_map(
+							function( $segment_id ) {
+								return Newspack_Popups_Segmentation::get_segment( $segment_id );
```

Comment:
This might result in `null` if a segment was deleted - this array should be filtered (or reduce used instead of map)

Reply:
Good catch—updated in cde98b8.